File: Java/백준/Silver/2630. 색종이 만들기/색종이 만들기.java
import java.io.*;
import java.util.*;

public class Main {

    static int[][] arr;
    static int white = 0;
    static int blue = 0;
    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        arr = new int[N][N];

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        solution(0, 0, N);

        System.out.println(white);
        System.out.println(blue);
    }

    static void solution(int r, int c, int size) {

        if (check(r, c, size)) {
            if (arr[r][c] == 1) {
                blue++;
            } else {
                white++;
            }
            return;
        }

        size /= 2;
        solution(r, c, size);
        solution(r + size, c, size);
        solution(r, c + size, size);
        solution(r + size, c + size, size);
    }

    static boolean check(int r, int c, int size) {

        int color = arr[r][c];

        for (int i = r; i < r + size; i++) {
            for (int j = c; j < c + size; j++) {
                if (arr[i][j] != color) {
                    return false;
                }
            }
        }

        return true;
    }
}

File: Python/백준/Silver/2630. 색종이 만들기/색종이 만들기.py
def check(r, c, size):
    global white, blue
    cnt = 0
    for i in range(r, r + size):
        for j in range(c, c + size):
            cnt += arr[i][j]
    if cnt == 0:
        white += 1
        return True
    elif cnt == size * size:
        blue += 1
        return True

    return False


def solution(r, c, size):
    if check(r, c, size):
        return

    solution(r, c, size // 2)
    solution(r + size // 2, c, size // 2)
    solution(r, c + size // 2, size // 2)
    solution(r + size // 2, c + size // 2, size // 2)


N = int(input())
arr = [list(map(int, input().split())) for _ in range(N)]

white = 0
blue = 0
solution(0, 0, N)

print(white)
print(blue)